File: app/src/main/java/org/stepic/droid/view/activities/RegisterActivity.java
package org.stepic.droid.view.activities;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import org.stepic.droid.R;
import org.stepic.droid.base.FragmentActivityBase;
import org.stepic.droid.core.ActivityFinisher;
import org.stepic.droid.core.ProgressHandler;
import org.stepic.droid.util.ProgressHelper;
import org.stepic.droid.util.ValidatorUtil;
import org.stepic.droid.web.RegistrationResponse;

import butterknife.Bind;
import butterknife.BindString;
import butterknife.ButterKnife;
import retrofit.Callback;
import retrofit.Response;
import retrofit.Retrofit;


public class RegisterActivity extends FragmentActivityBase {

    @Bind(R.id.root_view)
    View mRootView;

    @Bind(R.id.sign_up_btn)
    Button mCreateAccountButton;

    @Bind(R.id.actionbar_close_btn_layout)
    View mCloseButton;

    @Bind(R.id.first_name_reg)
    TextView mFirstNameView;

    @Bind(R.id.second_name_reg)
    TextView mSecondNameView;

    @Bind(R.id.email_reg)
    TextView mEmailView;

    @Bind(R.id.password_reg)
    TextView mPassword;

    @Bind(R.id.first_name_reg_wrapper)
    TextInputLayout mFirstNameViewWrapper;

    @Bind(R.id.second_name_reg_wrapper)
    TextInputLayout mSecondNameViewWrapper;

    @Bind(R.id.email_reg_wrapper)
    TextInputLayout mEmailViewWrapper;

    @Bind(R.id.password_reg_wrapper)
    TextInputLayout mPasswordWrapper;

    @BindString(R.string.password_too_short)
    String mPasswordTooShortMessage;


    ProgressDialog mProgress;
    TextWatcher mPasswordWatcher;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(org.stepic.droid.R.layout.activity_register);
        ButterKnife.bind(this);
        overridePendingTransition(org.stepic.droid.R.anim.slide_in_from_bottom, org.stepic.droid.R.anim.no_transition);

        hideSoftKeypad();

        mCreateAccountButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                createAccount();
            }
        });
        mCloseButton.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View v) {
                finish();
            }
        });

        mProgress = new ProgressDialog(this);
        mProgress.setTitle(getString(R.string.loading));
        mProgress.setMessage(getString(R.string.loading_message));
        mProgress.setCancelable(false);

        mPassword.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                boolean handled = false;
                if (actionId == EditorInfo.IME_ACTION_SEND) {
                    createAccount();
                    handled = true;
                }
                return handled;
            }
        });

        mPassword.addTextChangedListener(mPasswordWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (ValidatorUtil.isPasswordLengthValid(s.length())) {
                    hideError(mPasswordWrapper);
                }
            }
        });

        mRootView.requestFocus();
    }


    private void createAccount() {
        String firstName = mFirstNameView.getText().toString().trim();
        String lastName = mSecondNameView.getText().toString().trim();
        final String email = mEmailView.getText().toString().trim();
        final String password = mPassword.getText().toString();

        boolean isOk = true;

        if (!ValidatorUtil.isPasswordValid(password)) {
            showError(mPasswordWrapper, mPasswordTooShortMessage);
            isOk = false;
        }

        if (isOk) {
            hideError(mFirstNameViewWrapper);
            hideError(mSecondNameViewWrapper);
            hideError(mEmailViewWrapper);
            hideError(mPasswordWrapper);

            mShell.getApi().signUp(firstName, lastName, email, password).enqueue(new Callback<RegistrationResponse>() {
                @Override
                public void onResponse(Response<RegistrationResponse> response, Retrofit retrofit) {
                    ProgressHelper.dismiss(mProgress);
                    if (response.isSuccess()) {
                        mLoginManager.login(email, password, new ProgressHandler() {
                            @Override
                            public void activate() {
                                hideSoftKeypad();
                                ProgressHelper.activate(mProgress);
                            }

                            @Override
                            public void dismiss() {
                                ProgressHelper.dismiss(mProgress);
                            }
                        }, new ActivityFinisher() {
                            @Override
                            public void onFinish() {
                                finish();
                            }
                        });
                    } else {
                        Toast.makeText(RegisterActivity.this, "Failure " + response.code(), Toast.LENGTH_SHORT).show();
                    }

                }

                @Override
                public void onFailure(Throwable t) {
                    ProgressHelper.dismiss(mProgress);
                    Toast.makeText(RegisterActivity.this, "exception", Toast.LENGTH_SHORT).show();
                }
            });
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        bus.register(this);
    }

    @Override
    protected void onStop() {
        super.onStop();
        bus.unregister(this);
    }

    @Override
    protected void onDestroy() {
        mPassword.removeTextChangedListener(mPasswordWatcher);
        mPassword.setOnEditorActionListener(null);
        super.onDestroy();
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(org.stepic.droid.R.anim.no_transition, org.stepic.droid.R.anim.slide_out_to_bottom);
    }

    private void hideError(TextInputLayout textInputLayout) {
        if (textInputLayout != null) {
            textInputLayout.setError("");
            textInputLayout.setErrorEnabled(false);
        }
    }

    private void showError(TextInputLayout textInputLayout, String errorText) {
        if (textInputLayout != null) {
            textInputLayout.setErrorEnabled(true);
            textInputLayout.setError(errorText);
        }
    }

}

File: app/src/main/java/org/stepik/android/view/auth/ui/activity/RegisterActivity.kt
package org.stepik.android.view.auth.ui.activity

import android.os.Bundle
import android.text.Editable
import android.text.Spannable
import android.text.SpannableString
import android.text.TextWatcher
import android.text.method.LinkMovementMethod
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.res.ResourcesCompat
import androidx.core.view.isVisible
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import kotlinx.android.synthetic.main.activity_register.*
import org.stepic.droid.R
import org.stepic.droid.analytic.Analytic
import org.stepic.droid.analytic.LoginInteractionType
import org.stepic.droid.base.App
import org.stepic.droid.ui.activities.SmartLockActivityBase
import org.stepic.droid.ui.dialogs.LoadingProgressDialogFragment
import org.stepic.droid.ui.util.setOnKeyboardOpenListener
import org.stepic.droid.ui.util.snackbar
import org.stepic.droid.util.ProgressHelper
import org.stepic.droid.util.ValidatorUtil
import org.stepic.droid.util.stripUnderlinesFromLinks
import org.stepic.droid.util.toBundle
import org.stepik.android.model.user.RegistrationCredentials
import org.stepik.android.presentation.auth.RegistrationPresenter
import org.stepik.android.presentation.auth.RegistrationView
import org.stepik.android.view.base.ui.span.TypefaceSpanCompat
import ru.nobird.android.view.base.ui.extension.hideKeyboard
import javax.inject.Inject

class RegisterActivity : SmartLockActivityBase(), RegistrationView {
    companion object {
        const val ERROR_DELIMITER = " "
        const val TAG = "RegisterActivity"

        init {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        }
    }

    @Inject
    internal lateinit var viewModelFactory: ViewModelProvider.Factory

    private lateinit var registrationPresenter: RegistrationPresenter

    private val passwordTooShortMessage by lazy {
        resources.getString(R.string.password_too_short)
    }
    private val termsMessageHtml by lazy {
        resources.getString(R.string.terms_message_register)
    }

    private val progressDialogFragment: DialogFragment =
        LoadingProgressDialogFragment.newInstance()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        injectComponent()
        registrationPresenter = ViewModelProviders
            .of(this, viewModelFactory)
            .get(RegistrationPresenter::class.java)

        initTitle()

        termsPrivacyRegisterTextView.movementMethod = LinkMovementMethod.getInstance()
        termsPrivacyRegisterTextView.text = textResolver.fromHtml(termsMessageHtml)
        stripUnderlinesFromLinks(termsPrivacyRegisterTextView)

        signUpButton.setOnClickListener { submit(LoginInteractionType.button) }

        passwordField.setOnEditorActionListener { _, actionId, _ ->
            var handled = false
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                analytic.reportEvent(Analytic.Registration.CLICK_SEND_IME)
                submit(LoginInteractionType.ime)
                handled = true
            }
            handled
        }
        val reportAnalyticWhenTextBecomeNotBlank = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
                if (s.isNullOrBlank()) {
                    analytic.reportEvent(Analytic.Registration.TYPING_TEXT_FIELDS)
                }
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(s: Editable?) {
                registrationPresenter.onFormChanged()
                setSignUpButtonState()
            }
        }

        firstNameField.addTextChangedListener(reportAnalyticWhenTextBecomeNotBlank)
        emailField.addTextChangedListener(reportAnalyticWhenTextBecomeNotBlank)
        passwordField.addTextChangedListener(reportAnalyticWhenTextBecomeNotBlank)

        val onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
            if (hasFocus) {
                analytic.reportEvent(Analytic.Registration.TAP_ON_FIELDS)
            }
        }
        firstNameField.onFocusChangeListener = onFocusChangeListener
        emailField.onFocusChangeListener = onFocusChangeListener
        passwordField.onFocusChangeListener = onFocusChangeListener

        firstNameField.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_NEXT) {
                emailField.requestFocus()
                true
            } else {
                false
            }
        }

        emailField.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_NEXT) {
                passwordField.requestFocus()
                true
            } else {
                false
            }
        }

        registerRootView.requestFocus()

        initGoogleApiClient()

        setSignUpButtonState()

        setOnKeyboardOpenListener(root_view, {
            stepikLogo.isVisible = false
            signUpText.isVisible = false
        }, {
            stepikLogo.isVisible = true
            signUpText.isVisible = true
        })

    }

    private fun injectComponent() {
        App.component()
            .authComponentBuilder()
            .build()
            .inject(this)
    }

    override fun onStart() {
        super.onStart()
        registrationPresenter.attachView(this)
    }

    override fun onStop() {
        registrationPresenter.detachView(this)
        super.onStop()
    }

    private fun setSignUpButtonState() {
        signUpButton.isEnabled = emailField.text.isNotBlank() && firstNameField.text.isNotBlank() && passwordField.text.isNotBlank()
    }

    private fun initTitle() {
        val signUpString = getString(R.string.sign_up)
        val signUpSuffix = getString(R.string.sign_up_with_email_suffix)

        val spannableSignIn = SpannableString(signUpString + signUpSuffix)
        val typeface = ResourcesCompat.getFont(this, R.font.roboto_medium)

        spannableSignIn.setSpan(TypefaceSpanCompat(typeface), 0, signUpString.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)

        signUpText.text = spannableSignIn
    }

    private fun submit(interactionType: LoginInteractionType) {
        analytic.reportEvent(Analytic.Registration.CLICK_WITH_INTERACTION_TYPE, interactionType.toBundle())
        currentFocus?.hideKeyboard()

        val firstName = firstNameField.text.toString().trim()
        val lastName = " " // registrationSecondName.text.toString().trim()
        val email = emailField.text.toString().trim()
        val password = passwordField.text.toString()

        analytic.reportEvent(Analytic.Interaction.CLICK_REGISTER_BUTTON)

        var isOk = true

        if (!ValidatorUtil.isPasswordValid(password)) {
            showError(passwordTooShortMessage) // todo
            isOk = false
        }

        if (isOk) {
            registrationPresenter.submit(RegistrationCredentials(firstName, lastName, email, password))
        }
    }

    override fun setState(state: RegistrationView.State) {
        if (state is RegistrationView.State.Loading) {
            ProgressHelper.activate(progressDialogFragment, supportFragmentManager, LoadingProgressDialogFragment.TAG)
        } else {
            ProgressHelper.dismiss(supportFragmentManager, LoadingProgressDialogFragment.TAG)
        }

        when (state) {
            is RegistrationView.State.Idle -> {
                signUpButton.isEnabled = true
                registerForm.isEnabled = true
                registerErrorMessage.isVisible = false
            }

            is RegistrationView.State.Error -> {
                showError(getErrorString(state.data.email))
                showError(getErrorString(state.data.firstName))
                showError(getErrorString(state.data.lastName))
                showError(getErrorString(state.data.password))
            }

            is RegistrationView.State.Success -> {
                screenManager.showLogin(this, state.credentials.login, state.credentials.password, true, courseFromExtra)
            }
        }
    }

    override fun showNetworkError() {
        registerRootView.snackbar(messageRes = R.string.connectionProblems)
    }

    override fun applyTransitionPrev() {} // we need default system animation

    private fun showError(errorText: String?) {
        errorText?.let {
            analytic.reportEventWithName(Analytic.Registration.ERROR, errorText)
            if (registerErrorMessage.visibility == View.GONE) {
                signUpButton.isEnabled = false
                registerForm.isEnabled = false
                registerErrorMessage.text = it
                registerErrorMessage.isVisible = true
            }
        }
    }

    private fun getErrorString(values: List<String?>?): String? =
        values
            ?.takeIf { it.isNotEmpty() }
            ?.joinToString(separator = ERROR_DELIMITER)
}